### PR TITLE
Better error handling for foundation resources and data sources

### PR DESCRIPTION
**Pull request by bhatipradeep**
Remove error handling from client level for some foundation entities and handle it from respective provider code level. 
Entities effected : 
- ipmi_config resource
- image_nodes resource
- node_network_details data source

### CODE REVIEW

File: client/foundation/foundation_structs.go [hunk 1]
```diff
@@ -163,14 +163,11 @@ type ImageNodesAPIResponse struct {
...
-type Details struct {
-}
-
 //Error details for image nodes errored response
 type Error struct {
-	Message   string  `json:"message"`
-	Details   Details `json:"details"`
-	SessionID string  `json:"session_id"`
+	Message   string            `json:"message"`
+	Details   map[string]string `json:"details"`
```
Comment (siddharth-kulshrestha) [inline client/foundation/foundation_structs.go:169]: Shouldn't Details be map[string]interface{}?

File: nutanix/resource_nutanix_foundation_image_nodes.go [hunk 2]
```diff
@@ -1170,3 +1176,41 @@ func expandBlocks(d *schema.ResourceData) ([]*foundation.Block, error) {
...
+	var diags diag.Diagnostics
+
+	// append errors for failed node imaging
+	for _, v := range progress.Nodes {
+		if utils.Float64Value(v.TimeElapsed) < float64(AggregatePercentComplete) {
+			message := ""
+			for _, v1 := range v.Messages {
+				message += v1 + ". "
+			}
+			diags = append(diags, diag.Diagnostic{
```
Comment (siddharth-kulshrestha) [inline nutanix/resource_nutanix_foundation_image_nodes.go:1192]: Can't we directly use `diag.Errorf()`

File: nutanix/resource_nutanix_foundation_image_nodes.go [hunk 3]
```diff
@@ -1170,3 +1176,41 @@ func expandBlocks(d *schema.ResourceData) ([]*foundation.Block, error) {
...
+	}
+
+	// append errors for failed cluster creation
+	for _, v := range progress.Clusters {
+		if utils.Float64Value(v.TimeElapsed) < float64(AggregatePercentComplete) {
+			message := ""
+			for _, v1 := range v.Messages {
+				message += v1 + ". "
+			}
+			diags = append(diags, diag.Diagnostic{
```
Comment (siddharth-kulshrestha) [inline nutanix/resource_nutanix_foundation_image_nodes.go:1207]: Can't we directly use diag.Errorf() ?

File: nutanix/resource_nutanix_foundation_ipmi_config.go [hunk 4]
```diff
@@ -193,11 +210,9 @@ func resourceFoundationIPMIConfigCreate(ctx context.Context, d *schema.ResourceD
 		nodes := make([]map[string]interface{}, len(v.Nodes))
 		for k1, v1 := range v.Nodes {
 			node := make(map[string]interface{})
-			node["ipmi_configure_successful"] = v1.IpmiConfigureSuccessful
-			node["ipmi_configure_now"] = v1.IpmiConfigureNow
+			node["ipmi_configure_now"] = true
 			node["ipmi_ip"] = v1.IpmiIP
 			node["ipmi_mac"] = v1.IpmiMac
-			node["ipmi_message"] = v1.IpmiMessage
```
Comment (siddharth-kulshrestha) [inline nutanix/resource_nutanix_foundation_ipmi_config.go:200]: Why are we removing this ? 
Have we also removed this from schema ?

File: nutanix/resource_nutanix_foundation_ipmi_config.go [hunk 5]
```diff
@@ -170,6 +169,24 @@ func resourceFoundationIPMIConfigCreate(ctx context.Context, d *schema.ResourceD
 		return diag.FromErr(err)
 	}
 
+	// incase of errored response
+	if resp.Error != nil {
+		// check if error details exists for every ipmi IP
+		var diags diag.Diagnostics
+		for k, v := range resp.Error.Details {
+			diags = append(diags, diag.Diagnostic{
```
Comment (siddharth-kulshrestha) [inline nutanix/resource_nutanix_foundation_ipmi_config.go:177]: same as above

**Review by siddharth-kulshrestha — CHANGES_REQUESTED** [review]

File: client/foundation/foundation_structs.go (hunk 1 above)
Comment (bhati-pradeep) [reply client/foundation/foundation_structs.go:169]: Have checked with multiple scenarios. Details are always map of <string> : <string> or empty map

File: nutanix/resource_nutanix_foundation_image_nodes.go (hunk 2 above)
Comment (bhati-pradeep) [reply nutanix/resource_nutanix_foundation_image_nodes.go:1192]: diag.Errorf() returns diag.Diagnostics, but we need to create a diag.Diagnostic for every node failure and append them to create a diag.Diagnostics

File: nutanix/resource_nutanix_foundation_image_nodes.go (hunk 3 above)
Comment (bhati-pradeep) [reply nutanix/resource_nutanix_foundation_image_nodes.go:1207]: diag.Errorf() returns diag.Diagnostics, but we need to create a diag.Diagnostic for every cluster failure and append them to create a diag.Diagnostics

File: nutanix/resource_nutanix_foundation_ipmi_config.go (hunk 5 above)
Comment (bhati-pradeep) [reply nutanix/resource_nutanix_foundation_ipmi_config.go:177]: diag.Errorf() returns diag.Diagnostics, but we need to create a diag.Diagnostic for every ipmi ip config failure and append them to create a diag.Diagnostics

File: nutanix/resource_nutanix_foundation_ipmi_config.go (hunk 4 above)
Comment (bhati-pradeep) [reply nutanix/resource_nutanix_foundation_ipmi_config.go:200]: ipmi_message is not a part of model schema so I removed. Seems like I missed removing it from terraform schema. Let me fix it.

Comment (bhati-pradeep) [reply nutanix/resource_nutanix_foundation_ipmi_config.go:200]: @siddharth-kulshrestha Lets not remove them, as they are not affecting any flow.

**Pull request merged**